src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ObjectWriter.cs

```diff
@@ -166,15 +167,65 @@ public void EmitDebugLoc(int nativeOffset, int fileId, int linueNumber, int colN
         }
 
         [DllImport(NativeObjectWriterFileName)]
-        private static extern void FlushDebugLocs(IntPtr objWriter, string methodName, int methodSize);
-        public void FlushDebugLocs(int methodSize)
+        private static extern void EmitDebugVar(IntPtr objWriter, UInt32 varNumber, string name, UInt32 typeIndex, bool isParam, byte[] ranges, Int32 rangeCount);
+
+        public static byte[] GetBytes(NativeVarInfo[] nativeVarInfos)
+        {
+            int unitSize;
+            unsafe { unitSize = sizeof(NativeVarInfo); };
+            int len = nativeVarInfos.Length;
+            int size = unitSize * len;
+            byte[] bytes = new byte[size];
+
+            IntPtr ptr = Marshal.AllocHGlobal(unitSize);
+            for (int i = 0; i < len; i++)
+            {
+                Marshal.StructureToPtr(nativeVarInfos[i], ptr, true);
+                Marshal.Copy(ptr, bytes, unitSize * i, unitSize);
+            }
+            Marshal.FreeHGlobal(ptr);
+
+            return bytes;
+        }
+
+        public void EmitDebugVar(DebugVarInfo debugVar)
+        {
+            int rangeCount = debugVar.Ranges.Count;
+            int rangeSize;
+            unsafe { rangeSize = sizeof(NativeVarInfo); };
+            byte[] hackedRanges = GetBytes(debugVar.Ranges.ToArray());
+            EmitDebugVar(_nativeObjectWriter, debugVar.VarNumber, debugVar.Name, debugVar.TypeIndex, debugVar.IsParam, hackedRanges, rangeCount);
+        }
+
+        public void EmitDebugVarInfo(ObjectNode node)
+        {
+            // No interest if it's not a debug node.
+            if (!(node is INodeWithDebugInfo))
```

Comment:
Nit:
```
var nodeWithDebugInfo = node as INodeWithDebugInfo;
if (nodeWithDebugInfo != null)
```
Then you can also get rid of the `node as INodeWithDebugInfo` 4 lines down.